src/read/util.rs

```diff
@@ -0,0 +1,174 @@
+    type Storage;
+
+    const CAPACITY: usize;
+
+    fn as_slice(storage: &Self::Storage) -> &[MaybeUninit<Self::Item>];
+    fn as_mut_slice(storage: &mut Self::Storage) -> &mut [MaybeUninit<Self::Item>];
+}
+
+macro_rules! impl_array {
+    () => {};
+    ($n:literal $($rest:tt)*) => {
+        // SAFETY: `Storage` is `[MaybeUninit<Item>; CAPACITY]`.
+        unsafe impl<T> Array for [T; $n] {
+            type Item = T;
+            type Storage = [MaybeUninit<T>; $n];
+            const CAPACITY: usize = $n;
+
+            fn as_slice(storage: &Self::Storage) -> &[MaybeUninit<T>] {
+                storage
+            }
+
+            fn as_mut_slice(storage: &mut Self::Storage) -> &mut [MaybeUninit<T>] {
+                storage
+            }
+        }
+        impl_array!($($rest)*);
+    }
+}
+
+impl_array!(4 192);
+
+pub struct ArrayVec<A: Array> {
+    storage: A::Storage,
+    len: usize,
+}
+
+#[derive(Clone, Copy, Debug)]
+pub struct CapacityFull;
+
+impl<A: Array> ArrayVec<A> {
+    pub fn new() -> Self {
+        Self {
+            // SAFETY: An uninitialized `[MaybeUninit<_>; _]` is valid.
+            storage: unsafe { MaybeUninit::<A::Storage>::uninit().assume_init() },
+            len: 0,
+        }
+    }
+
+    pub fn as_ptr(&self) -> *const A::Item {
+        A::as_slice(&self.storage).as_ptr() as _
+    }
+
+    pub fn as_mut_ptr(&mut self) -> *mut A::Item {
+        A::as_mut_slice(&mut self.storage).as_mut_ptr() as _
+    }
+
+    pub fn clear(&mut self) {
+        let ptr: *mut [A::Item] = &mut **self;
+        // Set length first so the type invariant is upheld even if `drop_in_place` panicks.
+        self.len = 0;
+        // SAFETY: `ptr` contains valid elements only and we "forget" them by setting the length.
+        unsafe { ptr::drop_in_place(ptr) };
+    }
+
+    pub fn try_push(&mut self, value: A::Item) -> Result<(), CapacityFull> {
+        if self.len >= A::CAPACITY {
+            return Err(CapacityFull);
+        }
+
+        // SAFETY: This is essentially `MaybeUninit::write`, which is safe
+        // but only stablised in Rust 1.55 so we can't use due to MSRV.
+        unsafe {
+            A::as_mut_slice(&mut self.storage)[self.len]
+                .as_mut_ptr()
+                .write(value)
+        };
+        self.len += 1;
+        Ok(())
+    }
+
+    pub fn pop(&mut self) -> Option<A::Item> {
+        if self.len == 0 {
+            None
+        } else {
+            self.len -= 1;
+            // SAFETY: this element is valid and we "forget" it by setting the length.
+            unsafe { Some(ptr::read(self.as_ptr().add(self.len))) }
+        }
+    }
+
+    pub fn swap_remove(&mut self, index: usize) -> A::Item {
+        // Bounds check.
+        let _ = &self[index];
+        // SAFETY: replace `self[index]` with the last element.
+        // This works even if `self[index]` is the last element.
+        unsafe {
+            let last = ptr::read(self.as_ptr().add(self.len - 1));
```

Comment:
What if `self.len == 0`? I'd feel more comfortable if this function was implemented as `swap` + `pop`.

Reply:
This is how std implements `Vec::swap_remove`.

Reply:
Added a comment explaining `0 <= index < len`

Reply:
Switched to `swap_remove` anyway because it removes the usage of `unsafe`.